app/models/layer.js

```diff
@@ -0,0 +1,137 @@
+import Model from 'ember-data/model';
+import { computed, set } from '@ember/object';
+import { alias } from '@ember/object/computed';
+import { copy } from 'ember-copy';
+
+import { assign } from '@ember/polyfills';
+import { next } from '@ember/runloop';
+import DS from 'ember-data';
+
+const { attr, belongsTo } = DS;
+
+/**
+  Model for individual layers. Belongs to a layer-group. May be called individually for state changes.
+
+  @public
+  @class LayerModel
+*/
+export default Model.extend({
+  init(...args) {
+    this._super(...args);
+
+    // enforce presence of blank object for mapboxGL validation
+    if (!this.get('style.layout')) this.set('style.layout', {});
+
+    // determine which is the first occurring layer
+    // for testing, should check that a related layer group exists
+    if (this.get('layerGroup') && !this.get('layerGroup._firstOccurringLayer')) {
+      this.set('layerGroup._firstOccurringLayer', this.get('id'));
+      this.set('position', 1);
+    }
+
+    this.delegateVisibility();
+    this.addObserver('layerGroup.visible', this, 'delegateVisibility');
+  },
+
+  delegateVisibility() {
```

Comment:
I find if/elses hard to read, esp nested ones. I'm a fan of using returns to make the flow look nicer:
```
// Set visibility = false for singleton-type layers that are not at the top
if(this.get('layerVisibilityType') === 'singleton') {
  if(this.get('position') !== 1) {
    next(() => (!this.get('isDestroyed') ? this.set('visibility', false) : null));
    return;
  }
}
// set visibility = visible (from layerGroup) 
next(() => (!this.get('isDestroyed') ? this.set('visibility', visible) : null));
```